Create audio system

Verdanix opened the pull request:
# Feature Pull Request

## Feature Description
Using [Vosk](https://alphacephei.com/vosk/) and [Sherpa-onnx](https://github.com/k2-fsa/sherpa-onnx), I designed 2 interfaces and implementations.

Additionally, I designed a class for playing and transcribing audio

This will benefit developers and the Nora system, as playing/transcribing audio, TTS, and STT are kinda crucial

## Implementation Details
New files added:
- **SttEngine** allows developers to integrate their own STT engine
- **VoskSttEngine** is the implementation for Vosk
- **TtsEngine** allows for developers to integrate their own TTS engine
- **PiperTtsEngine** is the implementation for Sherpa-onnx
 
Modified files: 
- The root pom.xml

Key algorithms or techniques used
N/A

## API Changes
N/A

## Related Issues
N/A

## Checklist
- [x] Code follows the project's style guidelines
- [x] All new and existing tests pass
- [ ] New unit tests were added for the feature
- [ ] Integration tests cover the new functionality
- [x] Documentation was updated to reflect the new feature

Verdanix committed:
Update Pom.xml

Fixed
- Set Inception year to 2025 from 2024

Added
- Shade plugin
- Antrun plugin to download sherpa-onnx into /target/lib/sherpa-onnx.jar
- Maven install plugin to install local jar files into the maven local repository (.m2 directory)

Verdanix committed:
Updated root pom.xml

Fixed
- Sherpa-onnx not actually being installed by changing the goal from install to install-file

Verdanix committed:
Create Audio module

Added
- Pom.xml
- Testng test file

Verdanix committed:
Added AudioInterface and STT

Added
- Vosk STT support
- An class to handle read/write to speakers/migrophones

Verdanix committed:
Added/Updated Javadoc comments

Added
- Added Javadoc comments to VoskSttEngine and SttEngine
- Synchronized keyword to VoskSttEngine transcribe method
- Punctuation

Verdanix committed:
Added/Updated javadoc comments and created exceptions

Added
- Exceptions package with SttEngineInitializationException
- VoskInitializationException

Verdanix committed:
Reduced code duplication

Added
- SecurityUtils class which provides static methods for checking whether something's insecure
- throwIfFileIsInsecure method

Verdanix committed:
Updated javadocs and opening/closing of resources

Fixed
- Speaker and Mic only opens once
- Speaker and mic only open/close if they are open

Removed
- Irrelevant notes for the speak and transcribe methods
- finally statements as this is irrelevant

Verdanix committed:
Renamed Exceptions and reduced security (lil bit)

Fixed
- Exception naming to reduce number if classes (therefore reducing some memory)

Added
- throwIfIsNotDirectory to SecurityUtils
- @NonNull and final to throwIfFileIsInsecure

Removed
- throwIfFileIsInsecure call in VoskSttEngine. It now uses the new method, since the Vosk model isn't a file

Verdanix committed:
Made VoskSttEngine final

Verdanix committed:
Updated Javadocs and added TTS support

Fixed
- Irrelevant javadocs since it's inherited. Copilot committed this terrible mistake. I need somebody to blame ¯\_(ツ)_/¯ (kidding potential readers)

Added
- TTS support

I wasted like 88% of the time trying to debug an issue as to why my TTS was in a deeper voice. Turns out, the sample rate is to fault. I set it to 16000 when I was testing STT.

First I thought the model somehow corruptegd (Doesn't really make sense, but I was desperate). Then I thought it may be my USB speaker (then my computer speaker), then my headphones. Then I thought it was Linux driver (But realized it wouldn't be because VLC plays it normally). Then I thought it was IDEA. Then atlast, I **accidentally** clicked on a file that displayed the sample rate of the model (22050), and I cried really hard).

Verdanix committed:
Added package-info.java for TTS package

trammel-may commented on audio/src/main/java/dev/masonroot/audio/exceptions/PiperInitializationException.java:
```diff
@@ -0,0 +1,39 @@
+ *       information related to Piper initialization.
+ * </ul>
+ */
+public final class PiperInitializationException extends EngineInitializationException {
```
This exception should have a constructor which accepts the cause so that the stack trace contains the most amount of relevant information for debugging. Currently information is lost when this exception is being thrown as seen in `PiperTtsEngine` on line 60 where the causing `IOException` is erased.

trammel-may commented on audio/src/main/java/dev/masonroot/audio/exceptions/VoskInitializationException.java:
```diff
@@ -0,0 +1,39 @@
+ *       information related to Vosk initialization.
+ * </ul>
+ */
+public final class VoskInitializationException extends EngineInitializationException {
```
This exception should have a constructor which accepts the cause so that the stack trace contains the most amount of relevant information for debugging. Currently information is lost when this exception is being thrown as seen in VoskTtEngine on line 58 where the causing IOException is erased.

trammel-may commented on audio/src/main/java/dev/masonroot/audio/stt/VoskSttEngine.java:
```diff
@@ -0,0 +1,110 @@
+  }
+
+  @Override
+  public void close() throws Exception {
```
no exception is being thrown here, you can remove `throws Exception`

trammel-may commented on audio/src/main/java/dev/masonroot/audio/stt/VoskSttEngine.java:
```diff
@@ -0,0 +1,110 @@
+      this.model = new Model(modelPath.toString());
+      this.recognizer = new Recognizer(this.model, audio.microphone().getFormat().getSampleRate());
+    } catch (IOException e) {
+      NoraLogger.trace("Failed to load Vosk model.", e);
```
you should use the SLFj logger and not the NoraLogger. It is a static class wrapper for SLF4j and provides no useful abstraction

Verdanix replied:
The NoraLogger saves performance as said [here](https://github.com/Verdanix/Nora/pull/13#discussion_r1873487039)

> The class reduces startup time because the `@Slf4j` annotation takes around 400MS on that single annotation
> 
> Having one class is just better than way

trammel-may commented on audio/src/main/java/dev/masonroot/audio/AudioInterface.java:
```diff
@@ -0,0 +1,127 @@
+
+  @Override
+  public void close() throws Exception {
+    if (this.microphone.isOpen()) this.microphone.close();
+    if (this.speaker.isOpen()) this.speaker.close();
```
no need to check if its open, just call close its an idempotent call.

trammel-may commented on audio/src/main/java/dev/masonroot/audio/AudioInterface.java:
```diff
@@ -0,0 +1,127 @@
+   */
+  public synchronized byte[] read(final long timeoutInMs) {
+    final byte[] data = new byte[this.microphone.getBufferSize() / 5];
+    ByteArrayOutputStream out = new ByteArrayOutputStream();
```
can be final

Verdanix committed:
Removed irrelevant exception throwing

Verdanix committed:
Updated exception constructors

Added
Throwable parameter to EngineInitializationException children

Verdanix committed:
Added resource closing in TtsEngine and PiperTtsEngine

Verdanix committed:
Added Throwable parameter to EngineInitializationException

Verdanix committed:
Modified AudioInterface

Added
- The record is now final
- final to ByteArrayOutputStream

Removed
- isOpen checks in close metho

Verdanix committed:
Merge branch 'development' into create_audio_system
